File: android/app/src/main/java/com/rithvij/oclock/OClock.java
package com.rithvij.oclock;

import processing.core.*; 
import processing.data.*; 
import processing.event.*; 
import processing.opengl.*; 

import java.util.HashMap; 
import java.util.ArrayList; 
import java.io.File; 
import java.io.BufferedReader; 
import java.io.PrintWriter; 
import java.io.InputStream; 
import java.io.OutputStream; 
import java.io.IOException; 

public class OClock extends PApplet {

PImage oclock;
Gear[] gears;
Gear og;
Hand min;
Hand hour;
Hand sec;
boolean displayClock = true;
boolean displayImage = false;
boolean doUpdate = true;
boolean order = false;
boolean bounds = false;

// <enum>
final int HOUR = 312334;
final int MIN  = 321867;
final int SEC  = 318783;
//

final float MIN_HAND_LEN = 290;
final float SEC_HAND_LEN = 300;
final float HOUR_HAND_LEN = 180;

float CENTERX = 961, CENTERY = 540;
//final float CENTERX = 500, CENTERY = 500;

public void settings() {
  size(displayWidth, displayHeight, P2D);
}

public void setup () {
  //fullScreen();
  //surface.setResizable(true);
  oclock = loadImage("o'clock.png");

  initGears();
  initHands();
  setHands();
}

public void initGears() {
  //og = new OGear(0, 0, 54.0, 67, 18, "drawing");
  og = new OGear(0, 0, 54.0f, 67, 18, "ogears");
  gears = new Gear[9];
  gears[0] = new Gear(-12, -240, 43, 49, 8, "ninja");
  gears[1] = new Gear(-47.2f, -176.2f, 10, 20, 4, "plus");
  gears[2] = new Gear(-78.6f, -115.4f, 44, 55, 14, "four");
  gears[3] = new Gear(3.0f, -97.0f, 22, 36, 9, "pokemon");
  gears[4] = og;
  gears[5] = new Gear(-75.4f, 78, 30, 48, 12, "wheel");
  gears[6] = new Gear(61.8f, 56.4f, 13, 23, 6, "mine");
  gears[7] = new Gear(-29, 164, 38, 56, 15, "sharingan");
  gears[8] = new Gear(-98.2f, 209.6f, 19, 34, 9, "three");
}

public void initHands() {
  min = new Hand(og.pos, 272, PI/6, MIN);
  hour = new Hand(og.pos, 171, 5*PI/6, HOUR);
  sec = new Hand(og.pos, 272, -PI/2, SEC);
}

// clock
// r - 217 inner roman numeral touching circle
// r - 231 circle passing through center of X in "XII"
// r - 242 circle containing roman numerals inside
// r - 272 minute hand radius
// r - 171 small hand radius

// r - 295 outermost radius that contains the whole clock

public void draw() {
  // must have a background in draw to clear canvas
  background(0);
  //background(255, 204, 0);
  if (doUpdate) setHands();

  if (displayImage) {
    //println(oclock);
    //image(oclock, 0, -52);
    image(oclock, 2, -11.4f);
  }


  if (displayClock) {
    if (order) {
      drawHands();
    }
    for (int i=0; i< 9; i++) {
      gears[i].draw();
    }
    if (!order) {
      drawHands();
    }
  }
}

public void setHands() {
  min.value = minute();
  sec.value = second();
  hour.value = hour();
  //println(hour.value, min.value, sec.value);
}

public void drawHands() {
  min.draw();
  sec.draw();
  hour.draw();
}

public void keyPressed() {
  if (key == 'd') {
    //og.pos.x+=5;
    for (int i=0; i < 9; i++) {
      gears[i].pos.x += 5;
    }
    println("posx", og.pos.x);
  } else if (key == 'a') {
    //og.pos.x-=5;
    for (int i=0; i < 9; i++) {
      gears[i].pos.x -= 5;
    }
    println("posx", og.pos.x);
  }
  if (key == 'w') {
    //og.pos.y-=5;
    for (int i=0; i < 9; i++) {
      gears[i].pos.y -= 5;
    }
    println("posy", og.pos.y);
  } else if (key == 's') {
    //og.pos.y+=5;
    for (int i=0; i < 9; i++) {
      gears[i].pos.y += 5;
    }
    println("posy", og.pos.y);
  }
  if (key == 'o') {
    //og.iR ++;
    //og.oR ++;
    for (int i=0; i < 9; i++) {
      gears[i].oR += 1;
    }
    println(og.oR);
  } else if (key == 'p') {
    //og.iR --;
    //og.oR --;
    for (int i=0; i < 9; i++) {
      gears[i].oR -= 1;
    }
    println(og.oR);
  }

  if (key == ' ') displayClock = !displayClock;
  if (key == 'n') displayImage = !displayImage;
  if (key == 'v') doUpdate = !doUpdate;
  if (key == 'm') order = !order;
  if (key == 'b') bounds = !bounds;

  //https://forum.processing.org/two/discussion/comment/102455/#Comment_102455
  if ( key == 'p' ) looping = !looping;
  if ( key == 'r' && looping==false ) redraw();

  if (key == 'x') {
    saveFrame("images/line-######.png");
  }
}

public void mousePressed() {
  println(mouseX, mouseY);
}
// 949, 300 - topmost gear (s - 8) (innerR - 4, innerR - 43, innerSpikeR - 49, outerSpikeR - 57)
// 913.8, 363.8 - next small gear (s - 4) (innerR - 3, innerR - 10, outerSpikeR - 20) (scale - 0.62812495)
// 882.4, 424.6 - next gear 4 circle thing (s - 14) (innerR - 4, innerR - 44, outerSpikeR - 55)
// 4 circle thing small circles - (r - 14) (856.4, 424.2) centers seperation - 26
// 964, 443 - next (pokemon) (s - 9) (innerR - 4, outerR - 9, innerR - 22, outerSpikeR - 36)
// above center
// 961, 540 - center (s - 18)
// below orange
// 885.6, 618 - next left (s - 12) (innerR - 4, outerR - 15, innerR - 30, innerSpikeR - 38, outerSpikeR - 48)
// 1022.8, 596.4 - next right small (s - 6) (innerSpikeR - 13, outerSpikeR - 23))
// 932, 704 - next sharingan (s - 15) (innerR - 4, outerR - 17, innerR - 38, innerSpikeR - 49, outerSpikeR - 56)
// 862.8, 749.6 - last gear (s - 9) (innerR - 4, outerR - 19, innerSpikeR - 24, outerSpikeR - 34)
// last gear circles - (r - 5) (872, 758.2 - bottom right), (865.8, 738 - top), (850.8, 753.2 - bottom left)

class Gear {
  String name;
  PShape asset;
  int numSp = 6;
  float iR = 10;
  float oR = 30;
  PVector pos;
  PVector initPos;
  float imwidth, imheight;
  int grey = color(82, 82, 73);
  int boundcolor = color(255);
  Gear(float x, float y, float ir, float or, int spkes, String name) {
    println("new Gear");
    this.pos = new PVector(CENTERX + x, CENTERY + y);
    this.initPos = new PVector(x, y);
    //this.pos = new PVector(mouseX + x, mouseY + y);
    this.iR = ir;
    this.oR = or;
    this.numSp = spkes;
    this.name = name;
    this.asset = loadShape("assets/"+this.name+".svg");
    println("orig: name: "+name, this.asset.width, this.asset.height);
    this.imwidth = this.asset.width;
    this.imheight = this.asset.height;
    //println("upda: name: "+name, );
  }

  public void update() {
  }

  private void showBounds() {
    //pushMatrix();
    pushStyle();
    noFill();
    stroke(boundcolor);
    strokeWeight(10);
    rect(0, 0, imwidth, imheight);
    popStyle();
    //popMatrix();
  }

  public void draw() {
    CENTERX = width/2;
    CENTERY = height/2;
    this.pos.x = this.initPos.x + CENTERX;
    this.pos.y = this.initPos.y + CENTERY;

    pushMatrix();
    pushStyle();
    fill(grey);
    translate(this.pos.x, this.pos.y);
    //circle(0,0,this.oR * 2);
    //line(0, 0, mouseX-this.pos.x, mouseY-this.pos.y);
    popStyle();
    popMatrix();

    pushMatrix();
    pushStyle();
    //translate(mouseX + 240, mouseY - 240);
    float ang = map(mouseX, 0, width, 0, TWO_PI);
    //float scl = map(mouseY, 0, height, 0, 0.02);
    //println(this.oR * scl);

    translate(this.pos.x, this.pos.y);
    rotate(ang);
    //scale(scl);
    scale(this.oR * 0.0022037036f);
    //scale(this.oR * scl);

    translate(-this.asset.width/2, -this.asset.height/2);

    // these two
    //translate(49, -62.6);
    //scale(0.224);
    // are dependant

    shape(this.asset, 0, 0);

    if (bounds) showBounds();

    popStyle();
    popMatrix();
  }
}

class OGear extends Gear {
  int numSlices = 12;
  int orange = color(247, 144, 24);
  int yellow = color(255, 230, 99);
  OGear(float x, float y, float ir, float or, int spkes, String name) {
    super(x, y, ir, or, spkes, name);
    boundcolor = orange;
  }

  public void draw() {
    // r - 50 inner slices
    // r - 8  innermost light orange circle
    // r - 54 outer light orange circle
    // r - 67 outermost radius containing spikes inside
    // offsetAngle - 0.55850565 to align with image

    //push();
    //translate(this.pos.x, this.pos.y);
    ////float ang = map(mouseX, 0, width, 0, 2*PI);
    ////println(ang);
    ////rotate(ang);
    ////push();
    //translate(116, -149);
    //scale(0.535);
    //shape(this.asset, 0, 0);
    ////pop();
    ////shape(this.asset, 0, 0, mouseX, mouseY);
    //push();
    //noFill();
    //rect(0,0, imwidth, imheight);
    //pop();
    //pop();

    super.draw();
  }
}
class Hand {
  PVector start;
  float offAngle = PI / 180;
  float initAngle;
  float angle;
  float r;
  float value;
  PVector end;
  int type;
  Hand(PVector orange, float r, float angle, int type) {
    println("new Hand");
    this.start = orange;
    this.r = r;
    this.initAngle = angle;
    this.angle = angle;
    this.type = type;
    end = new PVector(this.start.x + this.r * cos(this.angle), this.start.y + this.r * sin(-this.angle));
  }

  public void update(float offsetAngle) {
    this.angle = this.initAngle + offsetAngle;
    end.x = this.start.x + this.r * cos(this.angle);
    end.y = this.start.y + this.r * sin(-this.angle);
  }
  public void update() {
    end.x = this.start.x + this.r * cos(this.angle);
    end.y = this.start.y + this.r * sin(-this.angle);
  }

  private void showBounds(float handLength) {
    pushStyle();
    noFill();
    stroke(255);
    strokeWeight(5);
    rect(0, -10, handLength, 20);
    popStyle();
  }

  public void draw() {
    float angle;
    if (this.type == MIN || this.type == SEC) {
      angle = map(this.value, 0, 60, 0, 2*PI);
    } else {
      angle = map(this.value % 12, 0, 12, 0, 2*PI);
    }

    pushMatrix();
    pushStyle();
    // disable Hand outline
    strokeWeight(0);
    //stroke(200);
    translate(this.start.x, this.start.y);
    rotate(angle - PI/2);
    if (this.type == MIN) {
      fill(48, 48, 42);
      triangle(0, 10, 0, -10, MIN_HAND_LEN, 0);

      if (bounds) showBounds(MIN_HAND_LEN);

      fill(0);
      ellipse(0, 0, 30, 30);
      //line(0, 0, MIN_HAND_LEN, 0);
    } else if (this.type == SEC) {
      fill(48, 48, 42);
      triangle(0, 10, 0, -10, SEC_HAND_LEN, 0);

      if (bounds) showBounds(SEC_HAND_LEN);

      fill(100);
      ellipse(0, 0, 20, 20);
      //line(0, 0, SEC_HAND_LEN, 0);
    } else if (this.type == HOUR) {
      fill(247, 100, 24);
      triangle(0, 10, 0, -10, HOUR_HAND_LEN, 0);

      if (bounds) showBounds(HOUR_HAND_LEN);

      fill(0);
      ellipse(0, 0, 10, 10);
      //line(0, 0, HOUR_HAND_LEN, 0);
    }
    popStyle();
    popMatrix();
  }
}
}

File: android/app/src/main/java/com/rithvij/oclock/OClock.kt
package com.rithvij.oclock

import androidx.preference.PreferenceManager
import processing.core.PApplet
import processing.core.PFont
import processing.core.PShape
import processing.core.PVector

class OClock : PApplet() {
    private lateinit var gears: Array<Gear>
    private lateinit var og: OGear

    private lateinit var min: Hand
    private lateinit var hour: Hand
    private lateinit var sec: Hand

    private lateinit var dial: Dial

    private val orange = color(247, 144, 24)

    //private int yellow = color(255, 230, 99);
    private val grey = color(82, 82, 73)
    private val handgrey = color(48, 48, 42)
    private val handorange = color(247, 100, 24)
    private val watermelon = color(214, 76, 83)
    private val tomato = color(255, 72, 0)
    private val kiwi = color(100, 195, 125)

    private var displayClock = true
    private var doUpdate = true
    private var order = false
    private var bounds = false

    // settings
    private var cycle = false
    private var duration = -1
    private var theme = "ogears"
    private var firstTime = true

    var version = "v0.0.4-alpha"

    // will be set to (displayWidth/2, displayHeight/2) in draw
    private var centerx = displayWidth / 2
    private var centery = displayHeight / 2

    private val minHandLen = 290f
    private val secHandLen = 300f
    private val hourHandLen = 180f

    override fun settings() {
        fullScreen(P2D)
    }

    override fun setup() {
        val prefs = PreferenceManager.getDefaultSharedPreferences(context)
        cycle = prefs.getBoolean("cycle", cycle)
        duration = prefs.getString("duration", "$duration")!!.toInt()
        theme = prefs.getString("theme", theme)!!
        firstTime = prefs.getBoolean("init", firstTime)

        kotlin.io.println("$cycle $duration $theme $firstTime")

        initGears()
        initHands()
        setHands()
        dial = Dial(og)
    }

    private fun initGears() {
        og = OGear(0f, 0f, 67f, 18, 1, arrayOf("ogears"))
        og.dialcolor = handorange
        gears = Array(9) { Gear(0f, 0f, 0f, 0, 1, arrayOf("")) }
        gears[0] = Gear(-12f, -240f, 55f, 8, 1, arrayOf("ninja"))
        gears[1] = Gear(-47.2f, -176.2f, 20f, 4, -1, arrayOf("plus"))
        gears[2] = Gear(-78.6f, -115.4f, 55f, 14, 1, arrayOf("four"))
        gears[3] = Gear(3.0f, -97.0f, 36f, 9, -1, arrayOf("pokemon"))
        gears[4] = og
        gears[5] = Gear(-75.4f, 78f, 48f, 12, -1, arrayOf("wheel"))
        gears[6] = Gear(61.8f, 56.4f, 23f, 6, -1, arrayOf("mine"))
        gears[7] = Gear(-29f, 164f, 56f, 15, 1, arrayOf("sharingan"))
        gears[8] = Gear(-98.2f, 209.6f, 34f, 9, -1, arrayOf("three"))
    }

    private fun initHands() {
        min = Hand(HandType.MIN)
        hour = Hand(HandType.HOUR)
        sec = Hand(HandType.SEC)
    }

    override fun draw() {
        // must have a background in draw to clear canvas
        background(0)
        if (doUpdate) setHands()
        if (displayClock) {
            translate(og.pos.x, og.pos.y)
            scale(2.02f)
            translate(-og.pos.x, -og.pos.y)
            dial.draw()
            if (order) {
                drawHands()
            }
            for (i in 0..8) {
                gears[i].draw()
            }
            if (!order) {
                drawHands()
            }
        }
    }

    private fun setHands() {
        min.value = minute().toFloat()
        sec.value = second().toFloat()
        hour.value = hour().toFloat()
    }

    private fun drawHands() {
        min.draw()
        sec.draw()
        hour.draw()
    }

    override fun mousePressed() {
//        println(mouseX, mouseY)
        if (mouseX > 3 * displayWidth / 4) {
            og.loadAssets(arrayOf("ogears"))
            og.dialcolor = handorange
        } else if (mouseX < displayWidth / 4) {
            og.loadAssets(arrayOf("watermelon"))
            og.dialcolor = watermelon
        } else if (mouseX > displayWidth / 4 && mouseX < 2 * displayWidth / 4) {
            og.loadAssets(arrayOf("kiwi"))
            og.dialcolor = kiwi
        } else if (mouseX > 2 * displayWidth / 4 && mouseX < 3 * displayWidth / 4) {
            og.loadAssets(arrayOf("tomato"))
            og.dialcolor = tomato
        }
        gears[4] = og
    }

    // roman numerals
    internal inner class Dial(var org: Gear) {
        private var mono: PFont = createFont("fonts/Trajan_Pro_Bold.ttf", 32f)

        init {
            textFont(mono)
            textAlign(CENTER)
        }

        fun draw() {
            pushMatrix()
            pushStyle()
            fill(grey)
            translate(this.org.pos.x, this.org.pos.y)
            for (i in 1..12) {
                if (i == 6 || i == 12) {
                    pushMatrix()
                    pushStyle()
                    rotate(i * TWO_PI / 12 - PI / 2)
                    fill(og.dialcolor)
                    ellipse(217f, 0f, 30f, 30f)
                    popStyle()
                    popMatrix()
                    continue
                }
                // deal with i == 7
                // deal with roman rotations
                pushMatrix()
                rotate(i * TWO_PI / 12 - PI / 2)
                translate(217f, 0f)
                rotate(PI / 2)
                translate(-217f, 0f)
                text(getRoman(i), 217f, 0f)
                popMatrix()
            }
            popStyle()
            popMatrix()
        }
    }


    private fun getRoman(number: Int): String {
        return when (number) {
            1 -> "I"
            2 -> "II"
            3 -> "III"
            4 -> "IV"
            5 -> "V"
            6 -> "VI"
            7 -> "VII"
            8 -> "VIII"
            9 -> "IX"
            10 -> "X"
            11 -> "XI"
            12 -> "XII"
            else -> "D"
        }
    }

    internal open inner class Gear(
        x: Float,
        y: Float,
        private var oR: Float,
        private var numSp: Int,
        private var direction: Int,
        namess: Array<String>
    ) {
        private var names: Array<String>
        private lateinit var assets: Array<PShape>
        private var driverTeeth = 18
        var pos: PVector = PVector(centerx + x, centery + y)
        private var initPos: PVector = PVector(x, y)
        private var imwidth: Float = 0f
        private var imheight: Float = 0f
        var boundcolor = color(255)
        var dialcolor = watermelon
        private var currangle = PI / 360

        init {
//            println("new Gear")
            this.names = Array(namess.size) { "" }
            for (i in namess.indices) {
                this.names[i] = namess[i]
            }
            this.loadAssets()
        }

        private fun loadAssets() {
            // to load or refresh assets
            this.assets = Array(this.names.size) { PShape() }
            for (i in this.assets.indices) {
//                kotlin.io.println("$i, ${this.names[i]}.svg")
                if (this.names[i].isNotEmpty())
                    this.assets[i] = loadShape("assets/" + this.names[i] + ".svg")
//                println("orig: name: " + this.names[i], this.assets[i].width, this.assets[i].height)
            }
            this.imwidth = this.assets[0].width
            this.imheight = this.assets[0].height
        }

        fun loadAssets(namess: Array<String>) {
//            val time_now = millis().toFloat()
            this.names = Array(namess.size) { "" }
            System.arraycopy(namess, 0, this.names, 0, namess.size)
            loadAssets()
//            println("time:", millis() - time_now)
        }

        private fun showBounds() {
            pushStyle()
            noFill()
            stroke(boundcolor)
            strokeWeight(10f)
            rect(0f, 0f, imwidth, imheight)
            popStyle()
        }

        fun draw() {
            centerx = (width / 2)
            centery = (height / 2)
            this.pos.x = this.initPos.x + centerx
            this.pos.y = this.initPos.y + centery

            pushMatrix()
            pushStyle()

            translate(this.pos.x, this.pos.y)
            this.currangle += PI / 360 * this.driverTeeth / this.numSp

            rotate(this.direction * this.currangle)
            scale(this.oR * 0.0022037036f)

            // translate to svg's center
            for (asset in this.assets) {
                pushMatrix()
                translate(-asset.width / 2, -asset.height / 2)
                shape(asset, 0f, 0f)
                popMatrix()
            }

            // debug bounds
            if (bounds) showBounds()

            popStyle()
            popMatrix()
        }
    }

    internal inner class OGear(
        x: Float,
        y: Float,
        or: Float,
        spikes: Int,
        d: Int,
        namess: Array<String>
    ) : Gear(x, y, or, spikes, d, namess) {
        init {
            boundcolor = orange
        }
    }

    internal enum class HandType {
        HOUR, MIN, SEC
    }

    internal inner class Hand(var type: HandType) {
        private var start: PVector = og.pos
        var value: Float = 0f
        private var angle: Float = 0f

        private fun showBounds(handLength: Float) {
            pushStyle()
            noFill()
            stroke(255)
            strokeWeight(5f)
            rect(0f, -10f, handLength, 20f)
            popStyle()
        }

        fun draw() {
            if (this.type == HandType.MIN || this.type == HandType.SEC) {
                this.angle = map(this.value, 0f, 60f, 0f, 2 * PI)
            } else {
                this.angle = map(this.value % 12, 0f, 12f, 0f, 2 * PI)
            }

            pushMatrix()
            pushStyle()
            // disable Hand outline
            strokeWeight(0f)
            translate(this.start.x, this.start.y)
            rotate(angle - PI / 2)
            when {
                this.type == HandType.MIN -> {
                    fill(handgrey)
                    triangle(0f, 10f, 0f, -10f, minHandLen, 0f)

                    if (bounds) showBounds(minHandLen)

                    fill(0)
                    ellipse(0f, 0f, 30f, 30f)
                }

                this.type == HandType.SEC -> {
                    fill(handgrey)
                    triangle(0f, 10f, 0f, -10f, secHandLen, 0f)

                    if (bounds) showBounds(secHandLen)

                    fill(100)
                    ellipse(0f, 0f, 20f, 20f)
                }

                this.type == HandType.HOUR -> {
                    fill(og.dialcolor)
                    triangle(0f, 10f, 0f, -10f, hourHandLen, 0f)

                    if (bounds) showBounds(hourHandLen)

                    fill(0)
                    ellipse(0f, 0f, 10f, 10f)
                }
            }
            popStyle()
            popMatrix()
        }
    }
}
